### ISSUE TITLE
localFile feature is broken

### ISSUE THREAD

**Issue by BarthV** [grafana/k6-operator#111]
Hi everyone !

I wanted to use scripts embedded in K6 runner containers but this feature seems only half-implemented.
Here's what I observed :
* verified against k8s `v1.24.0` & k8s `v1.22.2`
* using operator `ghcr.io/grafana/operator:controller-v0.0.7`
* `make install` / `make deploy` from git repo master (https://github.com/grafana/k6-operator/commit/9db5af67a7d026664eac5b2879db2103c96443cc)

```bash
> kubectl describe pods -n k6-operator-system k6-operator-controller-manager-b99d48f8b-ts895
  [...]
  manager:
    Container ID:  containerd://156c3dd7a9f3859d7116875871ab4ec747747c59ad6e1d5615e202a5a2eb80ad
    Image:         ghcr.io/grafana/operator:controller-v0.0.7
    Image ID:      ghcr.io/grafana/operator@sha256:b63818f86448a574a8c069b7089bcab4997569d396473e1b1b1f35fc8d7e9f40
    Port:          <none>
    Host Port:     <none>
    Command:
      /manager
    Args:
      --metrics-addr=127.0.0.1:8080
      --enable-leader-election
  [...]

> kubectl logs -n k6-operator-system k6-operator-controller-manager-b99d48f8b-ts895
2022-05-24T11:25:57.069Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "127.0.0.1:8080"}
2022-05-24T11:25:57.172Z	INFO	setup	starting manager
I0524 11:25:57.267031       1 leaderelection.go:242] attempting to acquire leader lease  k6-operator-system/fcdfce80.io...
2022-05-24T11:25:57.272Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
```

K6 custom resource
```yaml
---
apiVersion: k6.io/v1alpha1
kind: K6
metadata:
  name: k6-localfile-tst
spec:
  parallelism: 1
  script:
    localFile: /k6/tests/ci.js
  runner:
    image: registry.private.foo.bar/public/bundled-k6-runner:a2959ce
```

Reconcilliation log output :
```
2022-05-24T11:43:00.726Z	INFO	controllers.K6	Reconcile(); stage = initialized	{"k6": "k6-operator-system/k6-localfile-tst"}
2022-05-24T11:43:00.726Z	INFO	controllers.K6	Creating test jobs	{"k6": "k6-operator-system/k6-localfile-tst"}
2022-05-24T11:43:00.726Z	INFO	controllers.K6	Launching k6 test #1	{"k6": "k6-operator-system/k6-localfile-tst"}
2022-05-24T11:43:00.726Z	INFO	controllers.K6	Runner job is ready to start with image `registry.private.foo.bar/public/bundled-k6-runner:a2959ce` and command `[sh -c if [ ! -f /k6/tests/ci.js ]; then echo "LocalFile not found exiting..."; exit 1; fi;
k6 run --quiet /k6/tests/ci.js --address=0.0.0.0:6565 --paused]`	{"k6": "k6-operator-system/k6-localfile-tst"}
2022-05-24T11:43:00.735Z	ERROR	controllers.K6	Failed to launch k6 test	{"k6": "k6-operator-system/k6-localfile-tst", "error": "Job.batch \"k6-localfile-tst-1\" is invalid: [spec.template.spec.volumes[0].name: Required value, spec.template.spec.containers[0].volumeMounts[0].name: Required value, spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"\", spec.template.spec.containers[0].volumeMounts[0].mountPath: Required value]"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/grafana/k6-operator/controllers.launchTest
	/workspace/controllers/k6_create.go:80
github.com/grafana/k6-operator/controllers.createJobSpecs
	/workspace/controllers/k6_create.go:51
github.com/grafana/k6-operator/controllers.CreateJobs
	/workspace/controllers/k6_create.go:24
github.com/grafana/k6-operator/controllers.(*K6Reconciler).Reconcile
	/workspace/controllers/k6_controller.go:70
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:235
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
2022-05-24T11:43:00.735Z	ERROR	controller	Reconciler error	{"reconcilerGroup": "k6.io", "reconcilerKind": "K6", "controller": "k6", "name": "k6-localfile-tst", "namespace": "k6-operator-system", "error": "Job.batch \"k6-localfile-tst-1\" is invalid: [spec.template.spec.volumes[0].name: Required value, spec.template.spec.containers[0].volumeMounts[0].name: Required value, spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"\", spec.template.spec.containers[0].volumeMounts[0].mountPath: Required value]"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:237
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
```

Analysis :
* localFile feature has been added
* localFile doesn't require any Job defined volume & volumeMount to work
* Currently the controller keep trying to spawn runner jobs with empty volume, even when we don't need it :
    * https://github.com/grafana/k6-operator/blob/main/pkg/resources/jobs/initializer.go#L92-L103
    * https://github.com/grafana/k6-operator/blob/main/pkg/types/types.go#L87-L89
    * https://github.com/grafana/k6-operator/blob/main/pkg/resources/jobs/runner.go#L156-L164

Please tell me if I'm wrong but the current state of this feature seems totally broken.
We need to return an empty array of volume, not an array with a single empty default volume.

I successfully tested the localFile behavior when hard removing the volumes from the template, so I'm pretty confident that there's something wrong here. 
I'm seeing if I can work on some fix, but I you can see something better / faster I'm fully OK with that.

Thanks for your work

**Comment by guillermotti** [comment]
Hi @yorugac, we are facing the same problem.

Do you know when the PR will be merged? Thanks in advance!

**Comment by yorugac** [comment]
Done. New images were built. Please re-open if the issue repeats :slightly_smiling_face: Thank you!

**Comment by na--** [comment]
@yorugac, this (or something like it) still seems to be problem even with the latest `main` release, see https://community.k6.io/t/localfile-usage-throws-error-about-mount-volumes/4037/5

**Comment by jarviliam** [comment]
@yorugac I was dealing with the same issue a few days ago and following your advice on the k6 thread, I tried it again now with the latest tag set and it worked fine. From what I can see it seems like it was an issue in the previous latest image?
Anyways, thanks again for the update 👍

**Comment by yorugac** [comment]
@jarviliam Glad to hear that!

> From what I can see it seems like it was an issue in the previous latest image?

I think it was some cache problem. When I ran the [month-old image](ghcr.io/grafana/operator:6256831fbee63c8b73719a5dc6fc8b2f5a012a9d), local file feature worked correctly. So the image itself was fine; fix for regression was added back on May 25, https://github.com/grafana/k6-operator/commit/6256831fbee63c8b73719a5dc6fc8b2f5a012a9d and it's working since then. Hard to say what went off without records on what images exactly were used: e.g. did pull of `latest` return the real latest? Etc.
